src/target/adiv5_swd.c

```diff
@@ -79,133 +135,200 @@ static uint32_t firmware_dp_low_read(const uint16_t addr)
+		 * 
+		 * It is not possible to interrogate a multi-drop Serial Wire Debug system that includes multiple devices to establish
+		 * which devices are connected. Because all devices are selected on coming out of a line reset, no communication with
+		 * a device is possible without prior selection of that target using its target ID. Therefore, connection to a multi-drop
+		 * Serial Wire Debug system that includes multiple devices requires that either:
+		 * - The host has prior knowledge of the devices in the system and is configured before target connection.
+		 * - The host attempts auto-detection by issuing a target select command for each of the devices it has been
+		 * configured to support.
+	 	 */
+
+		/* Read DPIDR, if the first read fails, try the JTAG to SWD sequence, if that fails, give up */
+		uint32_t dpidr = 0;
+		bool tried_jtag_to_swd = false;
+		while (true) {
+			dpidr = adiv5_dp_read_dpidr(dp);
+			if (dpidr != 0)
+				/* Successfully read the DPIDR */
+				break;
+
+			if (!tried_jtag_to_swd) {
+				jtag_to_swd_sequence();
+
+				dp->fault = 0;
+
+				tried_jtag_to_swd = true;
+
+				/* Try again */
+				continue;
 			}
-		}
 
-		const uint8_t dp_version = (dp_dpidr & ADIV5_DP_DPIDR_VERSION_MASK) >> ADIV5_DP_DPIDR_VERSION_OFFSET;
-		if (dp_version >= 2) {
-			scan_multidrop = true;
-			// initial_dp must have the version field set so adiv5_dp_read() does protocol recovery correctly.
-			initial_dp->version = dp_version;
+			/* Give up */
+			DEBUG_ERROR("No usable DP found\n");
+			free(dp);
+			return 0;
+		}
 
+		/* DP must have the version field set so adiv5_dp_read() does protocol recovery correctly */
+		dp->version = (dpidr & ADIV5_DP_DPIDR_VERSION_MASK) >> ADIV5_DP_DPIDR_VERSION_OFFSET;
+		if (dp->version >= 2U) {
 			/* Read TargetID. Can be done with device in WFI, sleep or reset! */
 			/* TARGETID is on bank 2 */
-			adiv5_dp_write(initial_dp, ADIV5_DP_SELECT, ADIV5_DP_BANK2);
-			dp_targetid = adiv5_dp_read(initial_dp, ADIV5_DP_TARGETID);
-			adiv5_dp_write(initial_dp, ADIV5_DP_SELECT, ADIV5_DP_BANK0);
+			adiv5_dp_write(dp, ADIV5_DP_SELECT, ADIV5_DP_BANK2);
+			dp_targetid = adiv5_dp_read(dp, ADIV5_DP_TARGETID);
+			adiv5_dp_write(dp, ADIV5_DP_SELECT, ADIV5_DP_BANK0);
 		}
 	}
 
+	/* If we were given targetid or we have a DPv2+ device, do a multi-drop scan */
+#if PC_HOSTED == 0
+	/* On non hosted platforms, scan_multidrop can be constant */
+	const
+#endif
+		bool scan_multidrop = targetid || dp->version >= 2U;
+
 #if PC_HOSTED == 1
-	if (!initial_dp->dp_low_write) {
-		DEBUG_WARN("CMSIS_DAP < v1.2 can not handle multi-drop, disabling\n");
+	if (scan_multidrop && !dp->dp_low_write) {
+		DEBUG_WARN("Discovered multi-drop enabled target but CMSIS_DAP < v1.2 cannot handle multi-drop\n");
 		scan_multidrop = false;
 	}
 #endif
 
-	DEBUG_WARN("scan_multidrop: %s\n", scan_multidrop ? "true" : "false");
+	if (scan_multidrop)
+		adiv5_swdp_multidrop_scan(dp, dp_targetid);
+	else {
+		adiv5_dp_error(dp);
```

Comment:
This should be `adiv5_dp_abort(dp, ADIV5_DP_ABORT_STKERRCLR);` still from what we can see, to preserve the existing behaviour